client/go/outline/vpn.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package outline
+
+import (
+	"encoding/json"
+	"net"
+
+	perrs "github.com/Jigsaw-Code/outline-apps/client/go/outline/platerrors"
+	"github.com/Jigsaw-Code/outline-apps/client/go/outline/vpn"
+)
+
+type vpnConfigJSON struct {
+	VPNConfig       vpn.Config `json:"vpn"`
+	TransportConfig string     `json:"transport"`
+}
+
+// establishVPN establishes a VPN connection using the given configuration string.
+// The configuration string should be a JSON object containing the VPN configuration
+// and the transport configuration.
+//
+// The function returns a JSON string representing the established VPN connection,
+// or an error if the connection fails.
+func establishVPN(configStr string) (string, error) {
+	var conf vpnConfigJSON
+	if err := json.Unmarshal([]byte(configStr), &conf); err != nil {
+		return "", perrs.PlatformError{
+			Code:    perrs.IllegalConfig,
+			Message: "invalid VPN config format",
+			Cause:   perrs.ToPlatformError(err),
+		}
+	}
+
+	// Create Outline Client and Device
+	tcpControl, err := vpn.TCPDialerControl(&conf.VPNConfig)
+	if err != nil {
+		return "", err
+	}
+	tcp := net.Dialer{
+		Control:   tcpControl,
+		KeepAlive: -1,
+	}
+	udpControl, err := vpn.UDPDialerControl(&conf.VPNConfig)
+	if err != nil {
+		return "", err
+	}
+	udp := net.Dialer{Control: udpControl}
+	c, err := newClientWithBaseDialers(conf.TransportConfig, tcp, udp)
+	if err != nil {
+		return "", err
+	}
+	proxy, err := NewDevice(c)
+	if err != nil {
+		return "", err
+	}
+
+	// Establish system VPN to the proxy
+	conn, err := vpn.EstablishVPN(&conf.VPNConfig, proxy)
+	if err != nil {
+		return "", err
+	}
+
+	connJson, err := json.Marshal(conn)
```

Comment:
It doesn't seem like this json is that useful. The status will always be connected on success, and the ID is already an input. It seems we only need to return the status of the UDP support. Though we are not even using that. Perhaps we can just return an empty string, and we can think about how to track and display UDP status later.

Reply:
Right, as I mentioned above, this is not useful for this PR, but once we have network change notification these two fields (`ID`, `Status`) will be useful, we can remove the `SupportsUDP` though.

Reply:
Because it's not useful for the establishVPN API, let's remove it from this API. It's confusing, because the expectation is that it will be connected or not, based on the error, but your API is saying that it could end up in any state.

We can add it to a new API to monitor the state later.

And because we don't use in the UI, let's remove the UDP support too. It's misleading. If we want to show the UDP support, then must have the monitoring API. Let's design it properly when we get there.

Reply:
Deleted.